java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -142,16 +138,19 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
         case LESS_THAN:
         case LESS_THAN_OR_EQUAL_TO:
           symbolicValues = programState.peekValues(2);
-          setAsUndetermined(symbolicValues.get(1));
-          setAsUndetermined(symbolicValues.get(0));
+          removeZeroConstraint(symbolicValues.get(1));
+          removeZeroConstraint(symbolicValues.get(0));
           break;
         default:
           // do nothing
       }
     }
 
-    private void setAsUndetermined(SymbolicValue sv) {
-      programState = programState.addConstraint(sv, new ZeroConstraint(DivByZeroStatus.UNDETERMINED));
+    private void removeZeroConstraint(SymbolicValue sv) {
+      PMap<Class<? extends Constraint>, Constraint> constraints = programState.getConstraints(sv);
+      if(constraints != null) {
+        programState = programState.addConstraints(sv, constraints.remove(ZeroConstraint.class));
```

Comment:
Sounds like this will also help keep the PS cleaner and reduce increase of steps.